src/psyclone/psyad/adjoint_visitor.py

```diff
@@ -156,6 +158,41 @@ def schedule_node(self, node):
             else:
                 node_copy.children.append(result)
 
+        # Creating the adjoint may have altered the way variables are
+        # accessed within the code. If any of the active variables are
+        # subroutine arguments then we must update the access property of
+        # the associated ArgumentInterface.
+
+        # Since a piece of code could contain many Schedules, ensure we are
+        # currently handling the one representing the routine.
+        if isinstance(node, Routine):
+            dtools = DependencyTools()
+            # Input signatures ('in_sigs') are those whose first access is a
+            # read.
+            # Output signatures ('out_sigs') are those that are written to at
+            # some point.
+            in_sigs, out_sigs = dtools.get_in_out_parameters(
+                node_copy.children)
+            # Get the variable name associated with each of these signatures.
+            in_names = [sig.var_name for sig in in_sigs]
+            out_names = [sig.var_name for sig in out_sigs]
+
+            # We must update the symbols in the table of the new tree
+            adj_table = node_copy.symbol_table
+
+            for vname in self._active_variable_names:
+                sym = adj_table.lookup(vname)
+                if not sym.is_argument:
+                    continue
+                if vname in in_names:
+                    if vname in out_names:
+                        sym.interface.access = \
```

Comment:
It is here that I am concerned about potentially updating an interface that is share between symbols. I think we should copy the interface first to be safe.

Reply:
Ah, I see. Copy added along with a TODO.